pkg/dia/nft/nftTrade-scrapers/tofunft.go

```diff
@@ -0,0 +1,772 @@
+
+	log.Infof("processed %d trades", numTrades)
+
+	return nil
+}
+
+func (s *TofuNFTScraper) processTx(ctx context.Context, tx *utils.EthFilteredTx) (bool, error) {
+	log.Tracef("process tx -> block: %d, tx index: %d, tx hash: %s", tx.BlockNum, tx.TXIndex, tx.TXHash.Hex())
+
+	var ev tofunft.TofunftEvInventoryUpdate
+	err := tofuNFTABI.UnpackIntoInterface(&ev, "EvInventoryUpdate", tx.Logs[0].Data)
+	if err != nil {
+		log.Errorf("unable to read EvInventoryUpdate log from transaction(%s)", tx.TXHash)
+		return false, err
+	}
+
+	receipt, err := s.tradeScraper.ethConnection.TransactionReceipt(ctx, tx.TXHash)
+	if err != nil {
+		log.Errorf("unable to read transaction(%s) receipt: %s", tx.TXHash, err.Error())
+		return false, err
+	}
+
+	currSymbol := "ASTR"
+	currAddr := common.Address{}
+	currDecimals := 18
+
+	transfers, err := s.findERC721Transfers(ctx, receipt)
+	if err != nil {
+		log.Errorf("unable to find transfers of the event(block: %d, tx index: %d, tx: %s): %s", ev.Raw.BlockNumber, ev.Raw.TxIndex, ev.Raw.TxHash.Hex(), err.Error())
+		return false, err
+	}
+
+	// skip if the event has no transfer
+	if len(transfers) == 0 {
+		log.Tracef("event(block: %d, tx index: %d, tx: %s) skipped due to it has no erc721 transfer log", ev.Raw.BlockNumber, ev.Raw.TxIndex, ev.Raw.TxHash.Hex())
+		return true, nil
+	}
+
+	// skip if the event has multiple transfers due to we can't calculate the price of trade
+	if len(transfers) > 1 {
+		log.Tracef("event(block: %d, tx index: %d, tx: %s) skipped due to it has multiple erc721 transfer logs", ev.Raw.BlockNumber, ev.Raw.TxIndex, ev.Raw.TxHash.Hex())
+		return true, nil
+	}
+
+	normPrice := decimal.NewFromBigInt(ev.Inventory.Price, 0).Div(decimal.NewFromInt(10).Pow(decimal.NewFromInt(int64(currDecimals))))
+
+	usdPrice, err := s.calcUSDPrice(ev.Raw.BlockNumber, currAddr, currSymbol, normPrice)
+	if err != nil {
+		log.Errorf("unable to calculate usd price of the event(block: %d, log: %d, tx: %s): %s", ev.Raw.BlockNumber, ev.Raw.TxIndex, ev.Raw.TxHash.Hex(), err.Error())
+		return false, err
+	}
+
+	if err := s.notifyTrade(&ev, transfers[0], ev.Inventory.Price, normPrice, usdPrice, currSymbol, currAddr); err != nil {
+		if !errors.Is(err, errTofuNFTShutdownRequest) {
+			log.Warnf("event(block: %d, tx index: %d, tx: %s) couldn't processed: %s", ev.Raw.BlockNumber, ev.Raw.TxIndex, ev.Raw.TxHash.Hex(), err.Error())
+		}
+
+		return false, err
+	}
+
+	return false, nil
+}
+
+func (s *TofuNFTScraper) notifyTrade(ev *tofunft.TofunftEvInventoryUpdate, transfer *astarERC721Transfer, price *big.Int, priceDec decimal.Decimal, usdPrice float64, currSymbol string, currAddr common.Address) error {
+	nftClass, err := s.createOrReadNFTClass(transfer)
+	if err != nil {
+		return err
+	}
+
+	nft, err := s.createOrReadNFT(nftClass, transfer)
+	if err != nil {
+		return err
+	}
+
+	// Get block time.
+	timestamp, err := ethhelper.GetBlockTimeEth(int64(ev.Raw.BlockNumber), s.tradeScraper.datastore, s.tradeScraper.ethConnection)
```

Comment:
Please fetch the timestamp using s.tradeScraper.ethConnection.BlockByNumber(...) as this method is only applicable for Eth mainnet.